cpp/src/bitmask/bit_mask.h

```diff
@@ -0,0 +1,290 @@
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error destroy_bit_mask(bit_mask_t *valid) {
+    RMM_TRY(RMM_FREE(valid, 0));
+    return GDF_SUCCESS;
+  }
+
+  /**
+   *  @brief Get a single element of bits from the device
+   *
+   *  @param[out]  element - address on host where the bits will be stored
+   *  @param[out]  device_element - address on the device containing the bits to fetch
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error get_element(bit_mask_t *element, const bit_mask_t *device_element) {
+    CUDA_TRY(cudaMemcpy(element, device_element, sizeof(bit_mask_t), cudaMemcpyDeviceToHost));
+    return GDF_SUCCESS;
+  }
+	  
+  /**
+   *  @brief Put a single element of bits to the device
+   *
+   *  @param[out]  element - address on host containing bits to store
+   *  @param[out]  device_element - address on the device where the bits will be stored
+   *
+   *  @return GDF_SUCCESS on success, the CUDA error on failure
+   */
+  inline gdf_error put_element(bit_mask_t element, bit_mask_t *device_element) {
+    CUDA_TRY(cudaMemcpy(device_element, &element, sizeof(bit_mask_t), cudaMemcpyHostToDevice));
+    return GDF_SUCCESS;
+  }
+
+  /**
+   *  @brief Allocate device space for the valid bit mask.
+   *
+   *  @param[out] mask                  address of the bit mask pointer
+   *  @param[in]  number_of_records     number of records
+   *  @param[in]  fill_value            optional, should the memory be initialized to all 0 or 1s. All other
+   *                                    values indicate un-initialized.  Default is uninitialized
+   *  @param[in]  padding_bytes         optional, specifies byte boundary the data should be padded to.
+   *                                    Defaults to 64 bytes, meaning the space allocated will be rounded
+   *                                    up to the next multiple of 64 bytes.
+   *
+   *  @return GDF_SUCCESS on success, the RMM or CUDA error on error
+   */
+  gdf_error create_bit_mask(bit_mask_t **mask, gdf_size_type number_of_records, int fill_value = -1, gdf_size_type padding_bytes = 64) {
+    //
+    //  To handle padding, we will round the number_of_records up to the next padding boundary, then identify how many element
+    //  that equates to.  Then we can allocate the appropriate amount of storage.
+    //
+    gdf_size_type num_bytes = (number_of_records + 7) / 8;
+    gdf_size_type num_padding_blocks = (num_bytes + padding_bytes - 1) / padding_bytes;
+    gdf_size_type num_elements = bit_mask::num_elements(num_padding_blocks * 8 * padding_bytes);
+
+    RMM_TRY(RMM_ALLOC(mask, sizeof(bit_mask_t) * num_elements, 0));
+
+    if (fill_value == 0) {
+      CUDA_TRY(cudaMemset(*mask, 0, sizeof(bit_mask_t) * num_elements));
+    } else if (fill_value == 1) {
+      //
+      //  Filling with one is a more complex case, as we want the first
```

Comment:
Why does it matter that the padded elements have a value of 0 or 1? shouldn't the algoriothms operating on these things know that those positions are invalid anyway? Given that fill value can be defined by the user we provide no assurance of how this behaviour will be anyway so why not just set them all to 0xFF ?

Reply:
I agree. If people are writing code that makes assumptions about the bits outside of `[0, num_rows)`, then that is broken code. Therefore, no need to increase complexity of this code by ensuring only the bits `[0, num_rows)` are set. Just set all of them.

Reply:
Primary motivation for setting the last word correctly is for bit manipulation and counting.  If you have a bit mask referring to columns that are valid (the most common case I saw in cuDF) and you want a count of how many columns are valid, you can count how many ones are set.  If the size of your bit mask is not a multiple of 32 and the bits that pad you to a 32-bit boundary are not set to 0 then your count will be off.  Certainly the caller could be responsible for doing the right thing, but that becomes expensive and scattered through the code.  This code should be reasonably efficient and will allow the caller to assume that they can count 1s in the simple manner.

Reply:
I already have a kernel that counts the number of valid bits in a bitmask and I specifically made it such that it only counts the bits `[0, num_rows)`. https://github.com/rapidsai/cudf/blob/branch-0.5/cpp/src/bitmask/valid_ops.cu#L107

We definitely don't want to encourage people to rely on the bits outside `[0, num_rows)` being anything other than undefined.

Reply:
Agree that since we provide the counting code, we can simplify this code to treat the extra bits as undefined.